### PR TITLE
Add ability to scope posts by external model

### PR DESCRIPTION
**Pull request by paltman**
## Scoping

The idea of scoping allows you to setup your project to have multiple blogs
partitioned by whatever domain object you would like.

### Settings

* `PINAX_BLOG_SCOPING_MODEL` - a string in the format `"app.Model"` that will set a ForeignKey on the `blog.Post` model
* `PINAX_BLOG_SCOPING_URL_VAR` - the url variable name that you use in your url prefix that will allow you to look up your scoping object
* `PINAX_BLOG_HOOKSET` - introducing the hookset pattern from other apps.  just a single method: `get_scoped_object(self, **kwargs)` is defined.  override this in your project to get the object to scope your posts by.
* `pinax.blog.context_processors.scoped` - add to your context processors to put `scoped_var` in templates for url reversing

### Example

To demonstrate how to set all this up let's walk through an example where we
will scope by `auth.User` so that each user has their own blog at `/users/:username/`.

First we will modify the `settings.py`:

```python
# ... abbreviated for clarity

TEMPLATES = [
    {
        # ...
        "OPTIONS": {
            # ...
            "context_processors": [
                # ...
                "pinax.blog.context_processors.scoped"
            ],
        },
    },
]

PINAX_BLOG_SCOPING_URL_VAR = "username"
PINAX_BLOG_SCOPING_MODEL = "auth.User"
PINAX_BLOG_HOOKSET = "multiblog.hooks.HookSet"  # where `multiblog` is the package name of our project
```

Now, we'll add the url in `urls.py`:

```python
url(r"^users/(?P<username>[-\w]+)/", include("pinax.blog.urls", namespace="pinax_blog"))
```

And finally we'll implement our hookset by adding a `hooks.py`:

```python
from django.contrib.auth.models import User


class HookSet(object):

    def get_scoped_object(self, **kwargs):
        username = kwargs.get("username", None)
        return User.objects.get(username=username)
```

This is designed to work out of the box with templates in `pinax-theme-bootstrap`
so you can either use them directly or use them as a reference.  If you need to
reverse a URL for any of the `pinax-blog` urls you can simply do:

```django
{% url "pinax_blog:blog" scoped_var %}
```

Now that you have the context processor installed.

### CODE REVIEW

File: pinax/blog/models.py [hunk 1]
```diff
@@ -189,14 +193,22 @@ def save(self, **kwargs):
         super(Post, self).save(**kwargs)
 
     @property
+    def scoping_url_kwargs(self):
```
Comment (brosner) [inline pinax/blog/models.py]: There is an issue with this method only being available on `Post`. The scoper is now the "blog" object (due the absence of an actual `Blog` that would do the scoping). I suspect you will run into an issue in places where you are working with a list of posts.

File: pinax/blog/views.py [hunk 2]
```diff
@@ -164,6 +183,7 @@ def blog_feed(request, section=None, feed_type=None):
     else:
         raise Http404()
 
+    # @@@ how do we handle scoped kwarg here
```
Comment (brosner) [inline pinax/blog/views.py]: Ah, ha! My point earlier in the `Post` model was true and you noticed it too. To handle scoping here you need to reuse the code in `Post.scoping_url_kwargs`. Once that code is reusable outside of `Post` you can use the returned dictionary in `reverse("pinax_blog:blog")` below.

File: pinax/blog/context_processors.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from .conf import settings
+
+
+def scoped(request):
+    return {
+        "scoped_var": request.resolver_match.kwargs.get(settings.PINAX_BLOG_SCOPING_URL_VAR)
```
Comment (brosner) [inline pinax/blog/context_processors.py]: `scoped_var` is a misleading name. The value of this key is really the `scoper` lookup value.

File: pinax/blog/models.py [hunk 4]
```diff
@@ -61,6 +62,9 @@ class Post(models.Model):
 
     STATE_CHOICES = PINAX_BLOG_STATE_CHOICES
 
+    if settings.PINAX_BLOG_SCOPING_MODEL is not None:
+        scoped_for = models.ForeignKey(settings.PINAX_BLOG_SCOPING_MODEL, related_name="blog_posts")
```
Comment (brosner) [inline pinax/blog/models.py]: `scoped_for` and `scoper` seemed to be used interchangably throughout variable naming for this object. The former is a slightly confusing name. Reading code like `post.scoped_for` feels backwards to me. `post.scoper` would be better IMO.

File: pinax/blog/hooks.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+from pinax.blog.conf import settings
+
+
+class PinaxBlogDefaultHookSet(object):
+
+    def get_scoped_object(self, **kwargs):
```
Comment (brosner) [inline pinax/blog/hooks.py]: Another misleading name. The returned object from this method is the scoper and not the scoped object. The scoped object is the `Post` instance.

File: pinax/blog/models.py (hunk 1 above)
Comment (paltman) [inline pinax/blog/models.py]: Not sure I follow you completely.

File: pinax/blog/hooks.py (hunk 5 above)
Comment (paltman) [inline pinax/blog/hooks.py]: `scoper` it is!

File: pinax/blog/views.py (hunk 2 above)
Comment (paltman) [inline pinax/blog/views.py]: I think I see now.  Let me think on it.

Comment (paltman) [inline pinax/blog/views.py]: So i think this is solved now by having access to the kwargs (i changed the method sig but just didn't update the code)

File: pinax/blog/models.py (hunk 1 above)
Comment (paltman) [inline pinax/blog/models.py]: @brosner all fixed up but i still find this property useful on the `Post` object.

**Comment by jtauber** [comment]
In past discussions (blogs, wikis) we'd talked about the benefit of there being a single "container" model that is scoped, rather than scoping individual items (and you mention this in #77).

i.e. have an explicit `Blog` model and `Wiki` model and make *that* scoped (with posts and pages just having a foreign key to the `Blog` and `Wiki` respectively rather put the scoping on posts and pages).

**Comment by paltman** [comment]
Should I do that?  What would that give us?

**Comment by jtauber** [comment]
I think the main thing is it gives an explicit object that means "patrick's blog" or "james's blog for this course" or "this group's wiki" which is otherwise only implied. It's been a few years since I recall discussing it last. You said in #77 you thought it was needed, so you must have seen the need 11 months ago :-)

Some things that come to mind now though: 1) it gives a place to hang any configuration information (e.g. a configurable title for your blog, or preferences); 2) it seems like it would be useful if there is more than one object associated with the container (say you wanted Categories as well as Posts to be scoped).

**Comment by paltman** [comment]
@jtauber @brosner I like that.  In that case, I think the modeling should be `OneToOne` like:

```python
class Blog(models.Model):

    if settings.PINAX_BLOG_SCOPING_MODEL is not None:
        scoper = models.ForeignKey(settings.PINAX_BLOG_SCOPING_MODEL, related_name="blog")
```

So in the event of no `scoper`, it's just a `pk` until we decide to hang other information off of it.  I think for the case of `Section` I see pros and cons of it being scoped to `Blog` but also global so i'll just leave global for now.

Thoughts?

**Comment by paltman** [comment]
I guess going this route adds the additional complexity of needing to create the `Blog` object at some site defined trigger.

**Comment by paltman** [comment]
Ok, I solved that with an in-app receiver

**Pull request merged**